Fable.Remoting.DotnetClient/ClientRemoting.fs

```diff
@@ -0,0 +1,174 @@
+    }
+
+    // Warning: if routeBuilder returns a path with a leading /, it will override any subpath inside baseUrl
+    let createApi (baseUrl: string) =
+        {
+            RouteBuilder = None
+            BaseUri = Uri(baseUrl)
+            Client = None
+            AuthorizationToken = None
+            IsBinarySerialization = false
+            CustomHeaders = []
+        }
+
+    let withRouteBuilder (routeBuilder: string -> string -> string) options = { options with RouteBuilder = Some routeBuilder }
+
+    let withAuthorizationHeader token options = { options with AuthorizationToken = Some token }
+
+    let withBinarySerialization options = { options with IsBinarySerialization = true }
+
+    let withHttpClient client options = { options with Client = Some client }
+
+    let withCustomHeader (headers: (string * string) list) options = { options with CustomHeaders = headers @ options.CustomHeaders }
+
+    let buildProxy<'t> (options: RemoteBuilderOptions) : 't =
+        let isFSharpRecordType (t: Type) =
+            match t.GetCustomAttributes<CompilationMappingAttribute>() |> Seq.toList with
+            | [ attr ] -> attr.SourceConstructFlags.HasFlag SourceConstructFlags.RecordType
+            | _ -> false
+
+        let t = typeof<'t>
+        if not <| isFSharpRecordType t then failwithf "Type %s is not an F# record type" t.Name
+
+        let client = defaultArg options.Client (new HttpClient())
+        let routeBuilder = defaultArg options.RouteBuilder (fun x y -> sprintf "%s/%s" x y)
+
+        options.CustomHeaders |> List.iter (fun (name, value) -> client.DefaultRequestHeaders.Add(name, value))
+
+        match options.AuthorizationToken with
+        | Some token -> client.DefaultRequestHeaders.Add("Authorization", token)
+        | _ -> ()
+
+        let ctor = t.GetConstructors() |> Seq.exactlyOne
+        let parameters =
+            ctor.GetParameters()
+            |> Array.map (fun param ->
+                let funcType = param.ParameterType
+                let argTypes =
+                    Some funcType |> Seq.unfold (fun t ->
+                        match t with
+                        | Some t ->
+                            let generic = t.GetGenericTypeDefinition()
+                            if generic = typedefof<FSharpFunc<_,_>> then
+                                let args = t.GetGenericArguments()
+                                Some (args.[0], Some args.[1])
+                            elif generic = typedefof<Async<_>> then
+                                Some (t.GetGenericArguments().[0], None)
+                            else
+                                failwithf "Bad API record field %s, must be of type Async<'a> or a function returning Async<'a>" param.Name
+                        | None -> None
+                    )
+                    |> Seq.toArray
+
+                let route = Uri(options.BaseUri, routeBuilder t.Name param.Name) |> string
```

Comment:
> Warning: if routeBuilder returns a path with a leading /, it will override any subpath inside baseUrl

Maybe we can have some flexible string handling here to allow the `BaseUri` to end with `/` and the output of the `routeBuilder` to start with `/` with problems. I believe it goes something like this:
```fs
let baseUri = string options.BaseUri
let protocolRoute = routeBuilder t.Name param.Name
let route = baseUri.TrimEnd('/') + protocolRoute.TrimStart('/')
```
To be honest, not sure how `param.Name` becomes the function name of the protocol but I will have a deeper dive into it to understand the implementation

Reply:
Agreed, most times when a path starts with a leading /, it's not because someone wanted to override the subpath inside the base URL, but I like `Uri`'s handling of edge cases and errors. I think it'd be appropriate if we just strip the leading / from the route part.

> To be honest, not sure how param.Name becomes the function name of the protocol

F# compiles record types to sealed classes containing a single constructor, the names of the parameters to which are the names of the fields. So if you have `type MyType = { x: int; y: int }` the constructor would be `public MyType(int x, int y)`. Unless the record is marked as `CLIMutable`, which is an edge case I'll work around in the next commit.